subsys/mgmt/mcumgr/smp/src/mcumgr_client.c

```diff
@@ -0,0 +1,756 @@
+static K_MUTEX_DEFINE(mcumgr_mutex);
+
+static const char smp_images_str[] = "images";
+
+
+static bool get_echo_response(zcbor_state_t *state, struct mcumgr_os_echo_rsp *out)
+{
+	struct zcbor_string val = {0};
+	bool fin = zcbor_tstr_decode(state, &val);
+
+	if (!fin || !val.len) {
+		return false;
+	}
+
+	out->echo_res = val.value;
+	out->echo_rsp_len = val.len;
+	return true;
+}
+
+static int echo_res_fn(struct net_buf *nb, void *user_data)
+{
+	struct mcumgr_client_os_gr_rsp response;
+	zcbor_state_t zsd[4];
+	size_t decoded;
+	bool ok;
+	struct zcbor_map_decode_key_val echo_response[] = {
+		ZCBOR_MAP_DECODE_KEY_DECODER("r", get_echo_response, &response.echo)};
+
+	response.cmd = MGMT_OS_ECHO_RSP;
+	response.echo.echo_res = NULL;
+	response.echo.echo_rsp_len = 0;
+
+	if (!nb) {
+		LOG_ERR("Echo command timeout");
+		response.status = MGMT_ERR_ETIMEOUT;
+		client_gr_os_fn(&response);
+		k_sem_give(user_data);
+		return 0;
+	}
+
+	/* Init ZCOR decoder state */
+	zcbor_new_decode_state(zsd, ARRAY_SIZE(zsd), nb->data, nb->len, 1);
+
+	ok = zcbor_map_decode_bulk(zsd, echo_response, ARRAY_SIZE(echo_response), &decoded) == 0;
+
+	if (!ok) {
+		response.status = MGMT_ERR_ECORRUPT;
+		client_gr_os_fn(&response);
+		k_sem_give(user_data);
+		return MGMT_ERR_ECORRUPT;
+	}
+
+	response.status = MGMT_ERR_EOK;
+	client_gr_os_fn(&response);
+	k_sem_give(user_data);
+	return 0;
+}
+
+static int reset_res_fn(struct net_buf *nb, void *user_data)
+{
+	struct mcumgr_client_os_gr_rsp response;
+
+	if (!nb) {
+		response.status = MGMT_ERR_ETIMEOUT;
+	} else {
+		response.status = MGMT_ERR_EOK;
+	}
+	client_gr_os_fn(&response);
+	k_sem_give(user_data);
+	return 0;
+}
+
+static void reset_image_map_decoder_list(struct zcbor_map_decode_key_val *map, size_t map_size)
+{
+	for (int i = 0; i < map_size; i++) {
+		map[i].found = false;
+		map[i].value_ptr = &image_info[image_count];
+	}
+
+	image_info[image_count].active = false;
+	image_info[image_count].bootable = false;
+	image_info[image_count].confirmed = false;
+	image_info[image_count].pending = false;
+	image_info[image_count].permanent = false;
+	image_info[image_count].img_num = 0;
+	image_info[image_count].slot_num = -1;
+
+}
+
+static bool get_image_version(zcbor_state_t *state, struct mcumgr_image_data *out)
+{
+	struct zcbor_string val = {0};
+	bool fin = zcbor_tstr_decode(state, &val);
+
+	if (!fin || !val.len) {
+		return false;
+	}
+
+	if (val.len >= sizeof(out->version)) {
+		val.len = sizeof(out->version) - 1;
+	}
+	memcpy(out->version, val.value, val.len);
+	out->version[val.len] = '\0';
+	return true;
+}
+
+static bool get_image_hash(zcbor_state_t *state, struct mcumgr_image_data *out)
+{
+	struct zcbor_string val = {0};
+	bool fin = zcbor_bstr_decode(state, &val);
+
+	if (!fin || !val.len) {
+		return false;
+	}
+
+	if (val.len >= sizeof(out->hash)) {
+		val.len = sizeof(out->hash) - 1;
+	}
+	memcpy(out->hash, val.value, val.len);
+	out->hash[val.len] = '\0';
+	return true;
```

Comment:
Not really sure why any of this is needed when you can simply have the zcbor object itself and use it directly which avoids duplicating data because it doesn't copy the contents anywhere

Reply:
Now data is not duplicated. Code copy from message to image data buffer.